handler/rfc8693/handler.go

```diff
@@ -0,0 +1,268 @@
+	requestedTokenType := form.Get("requested_token_type")
+	switch requestedTokenType {
+	case tokenTypeAT, "":
+	default:
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("unsupported requested_token_type %s", requestedTokenType))
+	}
+
+	// From https://tools.ietf.org/html/rfc8693#section-2.1:
+	//
+	//	actor_token
+	//		OPTIONAL . A security token that represents the identity of the acting party.
+	//		Typically, this will be the party that is authorized to use the requested security
+	//		token and act on behalf of the subject.
+	actorToken := form.Get("actor_token")
+	if actorToken != "" {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("'actor_token' was provided but delegation is currently not supported."))
+	}
+
+	// From https://tools.ietf.org/html/rfc8693#section-2.1:
+	//
+	//	actor_token_type
+	//		An identifier, as described in Section 3, that indicates the type of the security token
+	//		in the actor_token parameter. This is REQUIRED when the actor_token parameter is present
+	//		in the request but MUST NOT be included otherwise.
+	actorTokenType := form.Get("actor_token_type")
+	if actorTokenType != "" {
+		return nil, errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("'actor_token_type' was provided but delegation is currently not supported."))
+	}
+
+	return &tokenExchangeParams{
+		subjectToken:     subjectToken,
+		subjectTokenType: subjectTokenType,
+	}, nil
+}
+
+func (c *Handler) HandleTokenEndpointRequest(ctx context.Context, requester fosite.AccessRequester) error {
+	if !c.CanHandleTokenEndpointRequest(requester) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	client := requester.GetClient()
+	if client.GetID() == "" {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("unauthenticated client"))
+	}
+
+	// Check whether client is allowed to use token exchange.
+	if !client.GetGrantTypes().Has(string(fosite.GrantTypeTokenExchange)) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHint("the client is not allowed to use token-exchange"))
+	}
+
+	// Get request parameter related token exchange.
+	params, err := parseRequestParameter(requester)
+	if err != nil {
+		return err
+	}
+
+	// Check and grant scope.
+	for _, scope := range requester.GetRequestedScopes() {
+		if !c.ScopeStrategy(client.GetScopes(), scope) {
+			return errorsx.WithStack(fosite.ErrInvalidScope.WithHintf("The OAuth 2.0 Client is not allowed to request scope '%s'.", scope))
+		}
+		requester.GrantScope(scope)
+	}
+
+	// Check and grant audience.
+	if err := c.AudienceMatchingStrategy(client.GetAudience(), requester.GetRequestedAudience()); err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("audience not match: %v", err))
+	}
+	for _, audience := range requester.GetRequestedAudience() {
+		requester.GrantAudience(audience)
+	}
+
+	// Verify subject token.
+	switch params.subjectTokenType {
```

Comment:
And an example for how `AccessTokenType` might be implemented:

```
func (c *AccessTokenType) Validate(ctx context.Context, config *TokenExchangeConfig, request fosite.AccessRequester, token string) (map[string]interface{}, error) {
	client := request.GetClient()

	sig := config.CoreStrategy.AccessTokenSignature(token)
	or, err := config.Storage.GetAccessTokenSession(ctx, sig, request.GetSession())
	if err != nil {
		return nil, errors.WithStack(fosite.ErrRequestUnauthorized.WithDebug(err.Error()))
	} else if err := config.CoreStrategy.ValidateAccessToken(ctx, or, token); err != nil {
		return nil, err
	}

	subjectTokenClientID := or.GetClient().GetID()
	// forbid original subjects client to exchange its own token
	if client.GetID() == subjectTokenClientID {
		return nil, errors.WithStack(fosite.ErrRequestForbidden.WithHint("Clients are not allowed to perform a token exchange on their own tokens"))
	}

	// Scope check
	for _, scope := range request.GetRequestedScopes() {
		if !config.ScopeStrategy(or.GetGrantedScopes(), scope) {
			return nil, errors.WithStack(fosite.ErrInvalidScope.WithHintf("The subject token is not granted \"%s\" and so this scope cannot be requested.", scope))
		}
	}

	// Convert to flat session with only access token claims
	var tokenObject map[string]interface{}
	if session, ok := or.GetSession().(Session); ok {
		tokenObject = session.AccessTokenClaimsMap()
		tokenObject["client_id"] = or.GetClient().GetID()
		tokenObject["scope"] = or.GetGrantedScopes()
		tokenObject["aud"] = or.GetGrantedAudience()
	} else {
		tokenObject = map[string]interface{}{
			"client_id": or.GetClient().GetID(),
			"scope":     or.GetGrantedScopes(),
			"sub":       or.GetSession().GetSubject(),
			"username":  or.GetSession().GetUsername(),
			"aud":       or.GetGrantedAudience(),
		}
	}

	return tokenObject, nil
}
```

Reply:
@vivshankar Thanks for the feedback. This makes it more flexible & extensible and can in a way support all the [subject token types](https://www.rfc-editor.org/rfc/rfc8693.html#section-3-2)

The only worrying point I have is that when implementing `AccessTokenType` the user will need to implement parts of the RFC that should in principle be provided in the library itself.

Reply:
My assumption when doing the implementation was that the support for the remaining token types will be added gradually. But I agree that this approach makes it hard to use other subject token types without modifying fosite.

Reply:
@saxenautkarsh We re-use the strategies, so this isn't actually duplicating the handlers outside of a couple of minor bits. I can submit a sample on my fork for you to look at, if interested, but it will take until at least this weekend.

Reply:
@vivshankar How about additionally we provide a default implementation of `AccessTokenType` interface? That will help in keeping it extensible for other token types and some custom implementations and at the same time easy to use because of the default implementation.

Reply:
@saxenautkarsh any news? this is a highly anticipated feature...

Reply:
> This is a highly anticipated feature...

@asv Thank you! Sorry for the delay, I will fix the above-mentioned issue this week.